util/resolver/pool.go

```diff
@@ -81,15 +83,37 @@ func (p *Pool) GetResolver(hosts docker.RegistryHosts, ref, scope string, sm *se
 		name = named.Name()
 	}
 
-	key := fmt.Sprintf("%s::%s", name, scope)
+	var key string
+	if strings.Contains(scope, "push") {
+		// When scope includes "push", index the authHandlerNS cache by session
+		// id(s) as well to prevent tokens with potential write access to third
+		// party registries from leaking between client sessions. The key will end
+		// up looking something like:
+		// 'wujskoey891qc5cv1edd3yj3p::repository:foo/bar::pull,push'
+		key = fmt.Sprintf("%s::%s::%s", strings.Join(session.AllSessionIDs(g), ":"), name, scope)
```

Comment:
What if this is just random, or something unique that is passed in from each call path that does push? `session.AllSessionIDs(g)` looks weird but I guess it has the correct behavior.

Reply:
I'm not sure I know enough about the codebase to say whether a random value would have the same exact effect. I would think use of the session ID would at least benefit sessions where multiple pushes are made for the same repositories. (e.g. cache exports?)